src/backend/distributed/transaction/remote_transaction.c

```diff
@@ -276,10 +276,22 @@ StartRemoteTransactionAbort(MultiConnection *connection)
 	Assert(transaction->transactionState != REMOTE_TRANS_INVALID);
 
 	/*
-	 * Clear previous results, so we have a better chance to send
-	 * ROLLBACK [PREPARED];
+	 * Clear previous results, so we have a better chance to send ROLLBACK
+	 * [PREPARED]. If we've previously sent a PREPARE TRANSACTION, we always
+	 * want to wait for that result, as that shouldn't take long and will
+	 * reserve resources.  But if there's another query running, we don't want
+	 * to wait, because a longrunning statement may be running, force it to be
+	 * killed in that case.
 	 */
-	ForgetResults(connection);
+	if (transaction->transactionState == REMOTE_TRANS_PREPARING ||
+		transaction->transactionState == REMOTE_TRANS_PREPARED)
+	{
+		ForgetResults(connection);
+	}
+	else if (!NonblockingForgetResults(connection))
+	{
+		ShutdownConnection(connection);
```

Comment:
Maybe do:
```
    if (transaction->transactionState == REMOTE_TRANS_PREPARING ||
       transaction->transactionState == REMOTE_TRANS_PREPARED)
    {
        StringInfoData command;

        /* await PREPARE TRANSACTION results, closing the connection would leave it dangling */
        ForgetResults(connection);

 	initStringInfo(&command);
 	appendStringInfo(&command, "ROLLBACK PREPARED '%s'",
                                   transaction->preparedName); 
        ...
    }
    else
    {
        if (!NonblockingForgetResults(connection))
        {
            /* can't sent ROLLBACK yet because worker is still busy, close the connection instead */
            ShutdownConnection(connection);
        }
        if (!SendRemoteCommand(connection, "ROLLBACK"))
        {
            ....
        }
        else
        {
            ...
        }
    }
```